sarc/core/scraping/users.py

```diff
@@ -0,0 +1,187 @@
+_user_scrapers = entry_points(group="sarc.user_scraper")
+
+
+def get_user_scraper(name: str) -> UserScraper:
+    """Raises KeyError if the name is not found"""
+    try:
+        return _builtin_scrapers[name]
+    except KeyError:
+        pass
+    val = _user_scrapers[name]
+    return val.load()()
+
+
+def update_user_match(*, value: UserMatch, update: UserMatch) -> None:
+    """
+    Fills in any missing information in value with the data in update.
+    """
+    if value.display_name is None:
+        value.display_name = update.display_name
+
+    if value.email is None:
+        value.email = update.email
+
+    name_dict = {mid.name: mid for mid in value.known_matches}
+
+    # Add the matching ids of the new usermatch to make sure that we have all
+    # the ids that this user is known under.
+    assert (
+        name_dict.get(update.matching_id.name, update.matching_id) == update.matching_id
+    )
+    value.known_matches.add(update.matching_id)
+    for mid in update.known_matches:
+        assert name_dict.get(mid.name, mid) == mid
+        value.known_matches.add(mid)
+
+    for domain, credentials in update.associated_accounts.items():
+        if domain not in value.associated_accounts:
+            value.associated_accounts[domain] = credentials
+        else:
+            value.associated_accounts[domain].merge_with(credentials, truncate=True)
+
+    value.supervisor.merge_with(update.supervisor, truncate=True)
+    value.co_supervisors.merge_with(update.co_supervisors, truncate=True)
+
+    value.github_username.merge_with(update.github_username, truncate=True)
+    value.google_scholar_profile.merge_with(
+        update.google_scholar_profile, truncate=True
+    )
+
+
+def scrape_users(
+    scrapers: list[tuple[str, Any]], _cache_policy: CachePolicy = CachePolicy.use
+) -> Iterable[UserMatch]:
+    """
+    Perform user scraping and matching according to the list of plugins passed in.
+
+    The first plugin to specify information wins in case of conflict.
+
+    This returns one UserMatch structure per scraped user, across all plugins.
+    The collected information is aggregated amongst plugins, but not with the
+    information in the database.
+    """
+    raw_data: dict[str, tuple[bytes, Any]] = {}
+    for scraper_name, config_data in scrapers:
+        try:
+            scraper = get_user_scraper(scraper_name)
+        except KeyError as e:
+            raise ValueError("Invalid user scraper") from e
+        config = scraper.validate_config(config_data)
+        raw_data[scraper_name] = (scraper.get_user_data(config), config)
+
+    # TODO: save the raw data for cache purposes
+    #  - Should refactor cache so that we don't *have* to use the function wrapper interface
+
+    # UserMatches, referenced by plugin name and matching id
+    user_refs: dict[MatchID, UserMatch] = {}
+    for scraper_name, (rdata, config) in raw_data.items():
+        for userm in scraper.parse_user_data(config, rdata):
```

Comment:
`scraper` here may be the latest scraper loaded in the previous loop, not the one corresponding to current `scraper_name` in this new loop

Reply:
You are absolutely right and I don't know why I didn't catch that before, fixed.